blueocean-core-js/src/js/ToastService.js

```diff
@@ -25,16 +25,24 @@ export class ToastService {
      * @returns {number} unique ID of toast
      */
     newToast(toast) {
-        if (!toast.id) {
-            toast.id = Math.random() * Math.pow(10, 16);
+        // prevent duplicate toasts from appearing when multiple UI elements
+        // are listening for an event that triggers creation of a toast
+        if (this._hasDuplicate(toast)) {
+            return null;
+        }
+
+        const newToast = toast;
+
+        if (!newToast.id) {
+            newToast.id = Math.random() * Math.pow(10, 16);
         }
 
         // TODO: determine why it's necessary to re-set the "toasts" field to trigger the UI update
-        const copy = this.toasts.slice();
-        copy.push(toast);
-        this.toasts = copy;
+        const toasts = this.toasts.slice();
```

Comment:
This definitely isn't a straight up MobX limitation: there is tons of sample code on the internet that modify the array in place (MobX actually uses its own ObservableArray under the hood). I just haven't had a chance to roll up sleeves and figure out what the problem is - yet.